File: projects/graph/src/storage/queue.js
/*
  1. Add a constructor with a storage structure; there are multiple options you could use for this
  2. Add a size getter that returns the number of items the queue is storing
  3. Add an `enqueue` method that accepts an item as input and adds it to the storage structure
  4. Add a `dequeue` method that removes the item in the queue that was added earliest
*/
/* eslint-disable no-throw-literal */
class Queue {
  constructor(arr = []) {
    this.storage = arr;
  }
  get copy() {
    return this.storage.slice();
  }
  get isEmpty() {
    return this.storage.length === 0;
  }
  get size() {
    return this.storage.length;
  }
  get top() {
    return this.storage[0];
  }
  forEach(f) {
    this.storage.forEach(e => f(e));
    return this;
  }
  enqueue(item) {
    this.storage.push(item);
  }
  dequeue() {
    if (this.storage.length) {
      return this.storage.shift();
    }
    return undefined;
  }
  clear() {
    this.storage = [];
    return this;
  }
}

module.exports = Queue;

File: projects/graph/backup/queue.py
class Queue:
    def __init__(self):
        self.storage = []
        self.count = 0

    def enqueue(self, item):
        self.storage.append(item)
        self.count += 1
        return
  
    def dequeue(self):
        if self.count > 0:
            self.count -= 1
            return self.storage.pop()
        else:
            return None

    def length(self):
        return self.count